xwiki-rendering-macros/xwiki-rendering-macro-figure/src/main/java/org/xwiki/rendering/internal/macro/figure/FigureMacro.java

```diff
@@ -81,16 +86,25 @@ public boolean supportsInlineMode()
     }
 
     @Override
-    public List<Block> execute(Object unusedParameters, String content, MacroTransformationContext context)
+    public List<Block> execute(FigureMacroParameters parameters, String content, MacroTransformationContext context)
         throws MacroExecutionException
     {
         XDOM xdom = this.contentParser.parse(content, context, false, false);
         // Mark the macro content as being content that has not been transformed (so that it can be edited inline).
         List<Block> contentBlock = List.of(new MetaDataBlock(xdom.getChildren(), getNonGeneratedContentMetaData()));
 
-        return List.of(new CompositeBlock(List.of(
-            new MacroBlock("figureTypeRecognizer", Map.of(), false),
-            new FigureBlock(contentBlock))
-        ));
+        FigureBlock figureBlock = new FigureBlock(contentBlock);
+        Block block;
+        // If a type is explicitly defined, it is used directly. Otherwise, we try to infer it from the macro's content.
+        if (parameters.getType() != null && !Objects.equals(parameters.getType(), FigureType.AUTOMATIC)) {
```

Comment:
I think the check for null is not needed according to https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Objects.html#equals(java.lang.Object,java.lang.Object)